### ISSUE TITLE
Adding parameters after update() gives assertion error

### ISSUE THREAD

**Issue by danielhers** [clab/dynet#889]
I'm trying to use DyNet in a scenario when not all parameters are known ahead of time (they depend in which kind of training examples I see), and some need to be added only after I've already completed the first batch. This very simple scenario fails:

    from dynet import *
    
    model = ParameterCollection()
    trainer = AdamTrainer(model)
    for _ in range(2):
        p = model.add_parameters((1,))
        x = parameter(p)
        x.forward()
        x.backward()
        trainer.update()

The error I'm getting is:

    eigen/unsupported/Eigen/CXX11/src/Tensor/TensorEvaluator.h:392: Eigen::TensorEvaluator<const Eigen::TensorCwiseBinaryOp<BinaryOp, LeftArgType, RightArgType>, Device>::TensorEvaluator(const XprType&, const Device&) [with BinaryOp = Eigen::internal::scalar_sum_op<float, float>; LeftArgType = const Eigen::TensorCwiseUnaryOp<Eigen::internal::bind2nd_op<Eigen::internal::scalar_product_op<float, float> >, const Eigen::TensorMap<Eigen::Tensor<float, 1> > >; RightArgType = const Eigen::TensorCwiseUnaryOp<Eigen::internal::bind2nd_op<Eigen::internal::scalar_product_op<float, float> >, const Eigen::TensorMap<Eigen::Tensor<float, 1> > >; Device = Eigen::DefaultDevice; Eigen::TensorEvaluator<const Eigen::TensorCwiseBinaryOp<BinaryOp, LeftArgType, RightArgType>, Device>::XprType = Eigen::TensorCwiseBinaryOp<Eigen::internal::scalar_sum_op<float, float>, const Eigen::TensorCwiseUnaryOp<Eigen::internal::bind2nd_op<Eigen::internal::scalar_product_op<float, float> >, const Eigen::TensorMap<Eigen::Tensor<float, 1> > >, const Eigen::TensorCwiseUnaryOp<Eigen::internal::bind2nd_op<Eigen::internal::scalar_product_op<float, float> >, const Eigen::TensorMap<Eigen::Tensor<float, 1> > > >]: Assertion `dimensions_match(m_leftImpl.dimensions(), m_rightImpl.dimensions())' failed.

Curiously, this works fine when I replace `AdamTrainer` with `SimpleSGDTrainer`.
Should what I'm trying to do even be possible with DyNet?

**Comment by pmichel31415** [comment]
I think this is because adam creates tensors to hold running averages of the moments when `trainer(model)` is called.

Not sure how to solve this easily, maybe the model should have access to the trainer...

**Comment by danielhers** [comment]
Can't the tensors just be resized when `trainer.update()` is called?

**Comment by pmichel31415** [comment]
Yes that would be an easy solution I think

**Comment by danielhers** [comment]
In the meantime, it would be nice if the trainers gave a meaningful error message when extra parameters have been added since the last `update()`. It would have saved me a lot of debugging time... :)

**Comment by danielhers** [comment]
Also, sometimes it's a `Segmentation fault` rather than an assertion error, which makes sense since we're accessing non-existent elements of an array.

**Comment by danielhers** [comment]
I'm definitely missing something, because this also happens with `SimpleSGDTrainer` if you repeat for enough iterations (change 2 to 100). However, in that case I'm not getting the Eigen assertion failure, but sometimes I'm getting `Segmentation fault` and sometimes this:

	free(): invalid next size (fast): 0x0000000001f97ce0 ***
	======= Backtrace: =========
	/lib/x86_64-linux-gnu/libc.so.6(+0x6ef45)[0x7f2d127caf45]
	/lib/x86_64-linux-gnu/libc.so.6(+0x746b6)[0x7f2d127d06b6]
	/lib/x86_64-linux-gnu/libc.so.6(+0x74e9e)[0x7f2d127d0e9e]
	/cs/usr/danielh/workspace/dynet/build/dynet/libdynet.so(_ZN5dynet26ParameterCollectionStorageD1Ev+0x1c8)[0x7f2d106340a8]
	/cs/usr/danielh/workspace/dynet/build/dynet/libdynet.so(_ZN5dynet19ParameterCollectionD1Ev+0x255)[0x7f2d10634ac5]
	/cs/nlp/danielh/workspace/tupa-dev/.virtualenv/lib/python3.5/site-packages/dyNET-2.0.2.dev1-py3.5-linux-x86_64.egg/_dynet.cpython-35m-x86_64-linux-gnu.so(+0x1f221)[0x7f2d121db221]
	/cs/nlp/danielh/workspace/tupa-dev/.virtualenv/bin/python[0x59aba2]
	/cs/nlp/danielh/workspace/tupa-dev/.virtualenv/bin/python[0x5b8fa3]
	/cs/nlp/danielh/workspace/tupa-dev/.virtualenv/bin/python(PyDict_SetItem+0x279)[0x599cb9]
	/cs/nlp/danielh/workspace/tupa-dev/.virtualenv/bin/python(PyImport_Cleanup+0x1ea)[0x523b3a]
	/cs/nlp/danielh/workspace/tupa-dev/.virtualenv/bin/python(Py_Finalize+0x61)[0x60f221]
	/cs/nlp/danielh/workspace/tupa-dev/.virtualenv/bin/python(Py_Main+0x641)[0x640e71]
	/cs/nlp/danielh/workspace/tupa-dev/.virtualenv/bin/python(main+0xe1)[0x4d0321]
	/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0)[0x7f2d1277c730]
	/cs/nlp/danielh/workspace/tupa-dev/.virtualenv/bin/python(_start+0x29)[0x5d8339]
	======= Memory map: ========
        ...

And sometimes this:

	munmap_chunk(): invalid pointer: 0x000000006859eda0 ***
	======= Backtrace: =========
	/lib/x86_64-linux-gnu/libc.so.6(+0x6ef45)[0x7ff1b68a1f45]
	/lib/x86_64-linux-gnu/libc.so.6(+0x746b6)[0x7ff1b68a76b6]
	/cs/usr/danielh/workspace/dynet/build/dynet/libdynet.so(_ZN5dynet16ComputationGraph5clearEv+0x16b)[0x7ff15bdce9ab]
	/cs/usr/danielh/workspace/dynet/build/dynet/libdynet.so(_ZN5dynet16ComputationGraphD1Ev+0x9)[0x7ff15bdcea09]
	/cs/nlp/danielh/workspace/tupa-dev/.virtualenv/lib/python3.5/site-packages/dyNET-2.0.2.dev1-py3.5-linux-x86_64.egg/_dynet.cpython-35m-x86_64-linux-gnu.so(+0x3d6f9)[0x7ff15da016f9]
	/cs/nlp/danielh/workspace/tupa-dev/.virtualenv/lib/python3.5/site-packages/dyNET-2.0.2.dev1-py3.5-linux-x86_64.egg/_dynet.cpython-35m-x86_64-linux-gnu.so(+0x24ab0)[0x7ff15d9e8ab0]
	python(PyCFunction_Call+0x77)[0x4e1cd7]
	python(PyEval_EvalFrameEx+0x48ca)[0x53ca8a]
	python(PyEval_EvalFrameEx+0x4b1f)[0x53ccdf]
	python[0x541b9f]
	python(PyEval_EvalFrameEx+0x5222)[0x53d3e2]
	python[0x541666]
	python(PyEval_EvalFrameEx+0x5222)[0x53d3e2]
	python[0x4e548c]
	python[0x5ae0c4]
	python[0x58b354]
	python(PyObject_Call+0x47)[0x5c8337]

This might explain why my PR #893 didn't fix the problem...

**Comment by pmichel31415** [comment]
Just to make sure, you are using ``dy.renew_cg()` in your script? It's not in the snippet you posted

**Comment by danielhers** [comment]
Yes, it happens with it too.

**Comment by danielhers** [comment]
I think I found the reason for the segfault, but there might still be more.
In [`ParameterCollectionStorage::gradient_l2_norm_dev`](https://github.com/clab/dynet/blob/master/dynet/model.cc#L785), the field `gradient_norm_scratch` is allocated once according to the size of the parameters list, and then never allocated again because it's no longer null. So if we calculate the gradient L2 norm again after adding parameters, we'll access unallocated memory.

**Comment by danielhers** [comment]
Still getting failures after my fix, but this time it's "floating point error". Backtrace:

    Program received signal SIGFPE, Arithmetic exception.
    0x00007fffc9bddeeb in dynet::LogSoftmax::aux_storage_size (this=0x6c4090c0) at /cs/usr/danielh/workspace/dynet_dev/dynet/nodes-softmaxes.cc:88
    88        return 2 * dim.size() / dim.rows() * sizeof(float);
    #0  0x00007fffc9bddeeb in dynet::LogSoftmax::aux_storage_size (this=0x6c4090c0) at /cs/usr/danielh/workspace/dynet_dev/dynet/nodes-softmaxes.cc:88
    #1  0x00007fffc98e9abd in dynet::BatchedExecutionEngine::incremental_forward_no_update (this=this@entry=0x369fce90, upto=upto@entry=89484, autobatch_strategy=<optimized out>) at /cs/usr/danielh/workspace/dynet_dev/dynet/exec.cc:632
    #2  0x00007fffc98f0f92 in dynet::BatchedExecutionEngine::incremental_forward (this=0x369fce90, i=89484) at /cs/usr/danielh/workspace/dynet_dev/dynet/exec.cc:793
    #3  0x00007fffcb50203e in __pyx_f_6_dynet_16ComputationGraph_inc_forward (__pyx_v_self=0x7fffdf22fd38, __pyx_v_index=89484, __pyx_skip_dispatch=<optimized out>) at _dynet.cpp:28143
    #4  0x00007fffcb51105e in __pyx_f_6_dynet_10Expression_forward (__pyx_v_self=0x7ffe7789f7f0, __pyx_skip_dispatch=<optimized out>, __pyx_optional_args=<optimized out>) at _dynet.cpp:34342
    #5  0x00007fffcb50093c in __pyx_pf_6_dynet_10Expression_22forward (__pyx_v_recalculate=<optimized out>, __pyx_v_self=0x7ffe7789f7f0) at _dynet.cpp:34440
    #6  __pyx_pw_6_dynet_10Expression_23forward (__pyx_v_self=0x7ffe7789f7f0, __pyx_args=<optimized out>, __pyx_kwds=0x0) at _dynet.cpp:34424
    #7  0x00000000004e1cd7 in PyCFunction_Call ()
